src/Bicep.Decompiler/Decompiler.cs

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Bicep.Core.Syntax;
+using System.IO;
+using Bicep.Core.PrettyPrint;
+using Bicep.Core.PrettyPrint.Options;
+using Bicep.Core.FileSystem;
+using System.Collections.Immutable;
+
+namespace Bicep.Decompiler
+{
+    public static class Decompiler
+    {
+        private static Uri ChangeExtension(Uri prevUri, string newExtension)
+        {
+            var uriString = prevUri.ToString();
+            var finalDot = uriString.LastIndexOf('.');
+            uriString = (finalDot >= 0 ? uriString.Substring(0, finalDot) : uriString) + $".{newExtension}";
+
+            return new Uri(uriString);
+        }
+
+        public static (Uri entrypointUri, ImmutableDictionary<Uri, string> filesToSave) DecompileFileWithModules(IFileResolver fileResolver, Uri jsonUri)
```

Comment:
I noticed that any failure we encounter results in an exception being thrown. Should we have the decompiler log errors just like the compiler does. If an issue is encountered that prevents decompilation, we could generate a commented out version of the original JSON or maybe some comments that indicate that we failed to process this part of the template.

Reply:
Good point, it would probably be more helpful to keep track of failures rather than throwing on the first.

Reply:
Btw, that shouldn't block this PR. (Although we should probably label the decompiler as experimental for people who install the latest all the time.)